### PR TITLE
core: Improve SEP-31 unit test coverage

### PR DESCRIPTION
**Pull request by lijamie98**
Improve the core's `org.stellar.anchor.sep31` package test coverage to 100%.

<details>
  <summary>PR Checklist</summary>

### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `paymentservice.stellar`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
</details>

### What

This PR is to improve SEP-31 core package test coverage to 100%. 

### Why

N/A

### Known limitations

This PR focuses on test coverage with some validations. More validation can be added to cover the corner cases.

### CODE REVIEW

**Comment by marcelosalloum** [comment]
I think we're missing the `@SerializedName("payments")` in JdbcSep31Refunds.refundPayments:
https://github.com/stellar/java-stellar-anchor-sdk/blob/f2e9f48b4642959081fe316fb88ad0f728f4c1bd/platform/src/main/java/org/stellar/anchor/platform/data/JdbcSep31Refunds.java#L10-L16

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt [hunk 1]
```diff
@@ -51,13 +49,13 @@ class Sep31ServiceTest {
     val gson: Gson = GsonUtils.getInstance()
 
     private const val stellarUSDC =
-      "stellar:USDC:GDQOE23CFSUMSVQK4Y5JHPPYK73VYCNHZHA7ENKCV37P6SUEO6XQBKPP"
+        "stellar:USDC:GDQOE23CFSUMSVQK4Y5JHPPYK73VYCNHZHA7ENKCV37P6SUEO6XQBKPP"
```
Comment (marcelosalloum) [inline core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt:52]: 👀  Can you re-format the code, so we don't add these whitespaces? You can do so by running `./gradlew spotlessApply`.

It should be automatically called in the pre-commit... not sure what happened 🤔

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt [hunk 2]
```diff
@@ -326,11 +351,84 @@ class Sep31ServiceTest {
...
-      txn.amountInAsset,
+        "stellar:USDC:GA5ZSEJYB37JRC5AVCIA5MOP4RHTM335X2KGX3IHOJAPP5RE34K4KZVN",
+        txn.amountInAsset,
     )
     assertEquals("12500", txn.amountOut)
     assertEquals(stellarJPYC, txn.amountOutAsset)
+
+    // Test null quote failure
+    Context.get().setQuote(null)
+    assertThrows<ServerErrorException> { sep31Service.updateTxAmountsBasedOnQuote() }
```
Comment (marcelosalloum) [inline core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt:362]: 💡  Can you assert the error class and message here? We do it in some places like:
https://github.com/stellar/java-stellar-anchor-sdk/blob/ece30f35d6fe37dfee2dab6df33791fd0d3ec470/core/src/test/kotlin/org/stellar/anchor/sep12/Sep12ServiceTest.kt#L49-L53

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt [hunk 3]
```diff
@@ -326,11 +351,84 @@ class Sep31ServiceTest {
...
+  }
+  @Test
+  fun test_getTransaction() {
+    assertThrows<BadRequestException> { sep31Service.getTransaction(null) }
+    assertThrows<BadRequestException> { sep31Service.getTransaction("") }
+
+    every { txnStore.findByTransactionId("not_found") } returns null
+    assertThrows<NotFoundException> { sep31Service.getTransaction("not_found") }
```
Comment (marcelosalloum) [inline core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt:370]: 💡 Can you assert the exception class and message here too?

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt [hunk 4]
```diff
@@ -326,11 +351,84 @@ class Sep31ServiceTest {
...
+    assertEquals(tr.amountFee, txn.amountFee)
+    assertEquals(tr.amountFeeAsset, txn.amountFeeAsset)
+    assertEquals(tr.stellarMemo, txn.stellarMemo)
+    assertEquals(tr.stellarMemoType, txn.stellarMemoType)
+    assertEquals(tr.completedAt, txn.completedAt)
+    assertEquals(tr.stellarTransactionId, txn.stellarTransactionId)
+    assertEquals(tr.externalTransactionId, txn.externalTransactionId)
+    assertEquals(tr.refunded, txn.refunded)
+    assertEquals(tr.requiredInfoMessage, txn.requiredInfoMessage)
+    assertEquals(tr.requiredInfoUpdates, txn.requiredInfoUpdates)
```
Comment (marcelosalloum) [inline core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt:390]: 👀  Why not assert the whole object in one line? Like `assertEquals(txn, gotTxn)`. Otherwise, the test won't fail when/if we add extra fields to the object schema.

Maybe you'll need to call `tr.id = txn.id` first.

Comment (marcelosalloum) [reply core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt:390]: Also, the order of the parameters seems inverted, the assertion methods usually take the expected value first and the actual value later, like in `assertEquals(wantTxn, gotTxn)`

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt [hunk 5]
```diff
@@ -715,35 +814,138 @@ class Sep31ServiceTest {
...
+    assertEquals(1000000, usdc.maxAmount)
+    assertNotNull(usdc.sep12)
+    assertNotNull(usdc.sep12.sender)
+    assertNotNull(usdc.sep12.sender.types)
+    assertEquals(3, usdc.sep12.sender.types.size)
+    assertNotNull(usdc.sep12.receiver)
+    assertNotNull(usdc.sep12.receiver.types)
+    assertEquals(2, usdc.sep12.receiver.types.size)
+    assertNotNull(usdc.fields.transaction)
+    assertEquals(3, usdc.fields.transaction.size)
```
Comment (marcelosalloum) [inline core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt:888]: 💡  As I said in a previous comment, I prefer comparing whole objects because it's easier to identify changes in the object schema that way.

When we compare individual fields like you're doing here, we don't catch inequalities in additional fields. Additional fields should break our tests for us to keep the tests up to date ✅

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt [hunk 6]
```diff
@@ -715,35 +814,138 @@ class Sep31ServiceTest {
...
+  @Test
+  fun test_updateFee() {
+    val jwtToken = TestHelper.createJwtToken()
+    Context.get().setRequest(request)
+    Context.get().setJwtToken(jwtToken)
+
+    // With quote
+    Context.get().setQuote(quote)
+    request.destinationAsset = "USDC"
+    sep31Service.updateFee()
```
Comment (marcelosalloum) [inline core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt:922]: 👀  It'd be nice to validate if the outcome of the `sep31Service.updateFee()` call is what we expect.

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt [hunk 7]
```diff
@@ -715,35 +814,138 @@ class Sep31ServiceTest {
...
+  @Test
+  fun test_updateFee_failure() {
+    val jwtToken = TestHelper.createJwtToken()
+    Context.get().setRequest(request)
+    Context.get().setJwtToken(jwtToken)
+
+    // With quote
+    Context.get().setQuote(quote)
+    quote.fee = null
+    assertThrows<SepValidationException> { sep31Service.updateFee() }
```
Comment (marcelosalloum) [inline core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt:949]: 💡  Can you validate the exception message being thrown here?

**Review by marcelosalloum — COMMENTED** [review]
Suuuuper cool to see our coverage reaching a higher level!

I've left a few suggestions below to improve the assertions and make sure the results are what we expect.

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31HelperTest.kt [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    txn.status = "pending_transaction_info_update"
+    Sep31Helper.validateStatus(txn)
+    txn.status = "pending_receiver"
+    Sep31Helper.validateStatus(txn)
+    txn.status = "pending_sender"
+    Sep31Helper.validateStatus(txn)
+    txn.status = "completed"
+    Sep31Helper.validateStatus(txn)
+    txn.status = "error"
+    Sep31Helper.validateStatus(txn)
```
[the remark this answers is not in the document]
Comment (lijamie98) [reply core/src/test/kotlin/org/stellar/anchor/sep31/Sep31HelperTest.kt]: I was trying to use the test code to lock the list of the `SepHelper.sep31Statuses` to prevent accidental modification of the list. 

I had revised the code to use `@ValueSource` to avoid the repetition.

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31HelperTest.kt [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+    txn.status = "Error"
+    assertThrows<BadRequestException> { Sep31Helper.validateStatus(txn) }
+    txn.status = "erroR"
+    assertThrows<BadRequestException> { Sep31Helper.validateStatus(txn) }
+    txn.status = ""
+    assertThrows<BadRequestException> { Sep31Helper.validateStatus(txn) }
+    txn.status = null
+    assertThrows<BadRequestException> { Sep31Helper.validateStatus(txn) }
```
[the remark this answers is not in the document]
Comment (lijamie98) [reply core/src/test/kotlin/org/stellar/anchor/sep31/Sep31HelperTest.kt]: 👍 I had revised the code as the following: 
```java
    val ex = assertThrows<BadRequestException> { Sep31Helper.validateStatus(txn) }
    assertEquals("'$status' is not a valid status of SEP31.", ex.message)
```

File: core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java [hunk 10]
```diff
@@ -106,7 +106,11 @@ public Sep31PostTransactionResponse postTransaction(
     Context.get().setAsset(assetInfo);
     Context.get().setTransactionFields(request.getFields().getTransaction());
 
+    // TODO: When there are missing required fields, we should catch Sep31MissingFieldException and
+    // set the status to pending_transaction_info_update
     validateRequiredFields();
+    // TODO: When there are missing customer infroamtion, we should catch
+    // Sep31CustomerInfoNeededException and the status to pending_customer_info_update
```
[the remark this answers is not in the document]
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java:113]: Updated. 😅

[the remark this answers is not in the document]
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java]: In the scenario of [this example](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0031.md#response), if `receiver_routing_number` field is missing in `postTransaction` call and we throw `Sep31MissingFieldException`, when do we set the transaction status to `pending_transaction_info_update`?

[the remark this answers is not in the document]
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java]: I had this question in mind when I was reviewing the `validateRequiredFields()`

[the remark this answers is not in the document]
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java]: If we do not have a clear answer at this moment, we can leave these TODO in the PR and review the code later.

File: core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java [hunk 11]
```diff
@@ -269,6 +276,12 @@ public Sep31GetTransactionResponse getTransaction(String id) throws AnchorExcept
       info("Unable to get transaction, 'id' is not provided");
       throw new BadRequestException("'id' is not provided");
     }
+
+    if (id.length() == 0) {
+      info("Empty transaction ID is not allowed");
+      throw new BadRequestException("'id' is empty");
+    }
+
```
[the remark this answers is not in the document]
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java:284]: Updated!

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt (hunk 3 above)
Comment (lijamie98) [reply core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt:370]: The exception class is checked by `assertThrows<NotFoundException>`.

Since there is only one place where `NotFoundException` is thrown, I skipped checking the message. I had added the message check so that we won't miss it if we add more `NotFoundException` to the function. 🙂

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt (hunk 4 above)
Comment (lijamie98) [reply core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt:391]: `tr` and `txn` are actually different types. 😢 

This is to make sure the fields of the `txn` returned from the datastore are copied correctly to the response.

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt (hunk 5 above)
Comment (lijamie98) [reply core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt:889]: Updated.

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt (hunk 7 above)
Comment (lijamie98) [reply core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt:929]: Updated.

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt (hunk 6 above)
Comment (lijamie98) [reply core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt:899]: Updated.

File: core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java [hunk 12]
```diff
@@ -280,6 +293,16 @@ public Sep31GetTransactionResponse getTransaction(String id) throws AnchorExcept
 
   public Sep31GetTransactionResponse patchTransaction(Sep31PatchTransactionRequest request)
       throws AnchorException {
+    if (request == null) {
+      infoF("request cannot be null");
+      throw new BadRequestException("request cannot be null");
+    }
+
+    if (request.getId() == null || request.getId().isEmpty()) {
```
[the remark this answers is not in the document]
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java:292]: Updated.

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt (hunk 2 above)
Comment (lijamie98) [reply core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt:359]: The type can actuallybe done by Java generics like assertThrows<ServerErrorException>.

I had added the message check.

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt (hunk 1 above)
Comment (lijamie98) [reply core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt:49]: Yeah... that's weird that the pre-commit hook did not happen. I will check later. 

`./gradlew spotlessApply` is now executed.

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31HelperTest.kt [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+  )
+  fun test_validateStatus(status: String) {
+    val txn = PojoSep31Transaction()
+    txn.status = status
+    Sep31Helper.validateStatus(txn)
+  }
+
+  @ParameterizedTest
+  @NullSource
+  @ValueSource(strings = ["Error", "erroR", ""])
```
Comment (marcelosalloum) [inline core/src/test/kotlin/org/stellar/anchor/sep31/Sep31HelperTest.kt:50]: Cool, I didn't know about `@NullSource`.

**Review by marcelosalloum — APPROVED** [review]
LGTM 👏
